Some adjustments to `watts_strogatz`

mlhetland opened the pull request:
Hi!

While using `watts_strogatz`, I noticed that it got stuck in an infinite loop for certain inputs. E.g.:

```julia
julia> watts_strogatz(11, 10, .2)
# hangs
```

I looked at the code to fix the bug, and noticed a few other minor issues that I thought I'd address. It might have been better to [open an issue first](https://github.com/JuliaGraphs/LightGraphs.jl/blob/b210395e8cbd109dc04c286f8af0c267cad47a85/CONTRIBUTING.md), before coding, but since I need a corrected version myself anyway, I though there was little potential for wasted effort, even if you decide not to go with my adjustments :-) I'll sketch out the rationale for the PR below.

**The infinite loop:** The original version uses trial and error to find a node to rewire to. If the source node is already connected to all other nodes, this leads to an infinite loop. This issue is handled by the condition `degree(g, s) < n - 1` for rewiring in my code, though even if this is removed, the reworked loop will be correct: Instead of trial and error, my version goes through the nodes (in random order) until a free one is found, rather than proceeding by trial and error. (I could have used reservoir sampling or something here, but `randperm` is very straightforward and obviously correct.) This ensures a single pass through the candidate nodes – but if the `degree(g, s) < n - 1` check were removed (if one wanted to do this), it could potentially lead to *no* valid neighbor being found. However, this is solved by permitting reattachment to the original neighbor (see below).

**Reattachment:** The original Watts–Strogatz paper is a bit vague on some details of their procedure, but e.g. Watts's book *Small Worlds* goes into a bit more detail. Specifically, he specifies deleting the edge in question before uniformly selecting a neighbor, excluding self-connections and multiple edges. In other words, it is valid to re-attach the edge to its original target, even if it's being rewired. This was not permitted in the original `watts_strogatz` function, affecting the probability distribution over the graphs.

**Always rewiring:** In the original, there was the following `if` statement:

```julia
if rand(rng) > β && !has_edge(g, s, target)
```

This triggered adding an edge (because it was not rewired). However, the `else` then was also triggered regardless of whether `rand(rng) > β`, as long as there was an edge there (e.g., rewired from somewhere else). In this case, it would automatically be rewired. This is no longer an issue in the new implementation. (Maybe this was part of the mechanism for getting the correct probability distribution, despite there not being edges there to block rewirings in the first place; if so, I'm just not understanding it properly :-))

**Two passes:** The original `watts_strogatz` function is optimized by not adding any edges initially, and considering each edge in turn, either adding it where it “should have been,” or where it would get rewired to. Now, it may be that the result is equivalent, but at least it's not obvious to me – and I think an obviously correct solution might be better, at least without an explanation of why the original was correct. The issue, it seems to me, is that when one considers rewiring in this version, many edges are missing, and will therefore not block potential rewiring sites. In my rewrite, I explicitly use the two phases consistently described by Watts and Strogatz in their various sources, first creating the lattice and then going through it, potentially rewiring. (I only delete the edge if it actually gets rewired to some other node than its original target, at least; and for a low `β` an performance hit from first adding and then rewiring should be pretty low.)

**Complete graph:** If `k == n - 1` and `k` is even, I just return a complete graph.

**Probability:** In Watts's description, he explicitly states that one should generate a random deviate `r`, and then use the conditions `r < β` and `r ≥ β` (i.e., not `r ≤ β` and `r > β`, as in the original `watts_strogats`). This is also used in the [NetworkX implementation](https://networkx.org/documentation/stable/_modules/networkx/generators/random_graphs.html#watts_strogatz_graph), by the way. Of courtse, our random value won't be from `[0,1]`, but `[0,1)`, but that does not make this less relevant. Specifically, it is consistent with never rewiring for `β = 0` and always rewiring for `β = 1`.

**Docstring:** I've also made some adjustments to the docstring (e.g., not stating that all nodes have degree `k`, but that they have an *expected* degree of `k` or `k - 1`, and adding a description of the procedure used).

mlhetland commented:
I see some checks weren't successful, but as far as I can see, they're not directly related to the PR? (All tests pass locally for me… :-})

sbromberger commented:
Thanks for this! Ignore the test failure on nightly. It's unrelated. I'll have some time to review this in the next few days.

sbromberger commented:
Also - could you post some benchmarks before/after this PR?

mlhetland commented:
> Also - could you post some benchmarks before/after this PR?

Absolutely! And I see I should have looked at that before, perhaps. Because there's a significant slowdown :-O

I didn't run full LightGraphs benchmarks (because I guess this stuff would drown a bit). Rather, I used this function:

```julia
function main()
    for n in [100, 500, 1000], k in 1:10, β in 0.0:.2:1.0
        watts_strogatz(n, k, β)
    end
    return nothing
end
```

Using `@benchmark` from BenchmarkTools, I got the following, with the original:

```
  memory estimate:  18.06 MiB
  allocs estimate:  248861
  --------------
  minimum time:     18.139 ms (0.00% GC)
  median time:      18.491 ms (0.00% GC)
  mean time:        19.488 ms (5.65% GC)
  maximum time:     22.136 ms (12.36% GC)
  --------------
  samples:          257
  evals/sample:     1
```

Now, with the new version:

```
  memory estimate:  757.44 MiB
  allocs estimate:  433989
  --------------
  minimum time:     1.434 s (5.64% GC)
  median time:      1.443 s (5.84% GC)
  mean time:        1.443 s (5.88% GC)
  maximum time:     1.451 s (6.03% GC)
  --------------
  samples:          4
  evals/sample:     1
```

One thing that's certainly wasteful is the allocation in `randperm`, so I changed that to an in-place version (where I just set up `collect(1:n)` initially, and then use `randperm!` in the loop), and that helped with the allocation, but it's still slow:

```
  memory estimate:  22.82 MiB
  allocs estimate:  314268
  --------------
  minimum time:     1.305 s (0.00% GC)
  median time:      1.307 s (0.11% GC)
  mean time:        1.308 s (0.11% GC)
  maximum time:     1.315 s (0.21% GC)
  --------------
  samples:          4
  evals/sample:     1
```

Of course, *some* slowdown is to be expected, as I removed some optimizations in order to make the procedure correct, but I'd still not expect a factor of ~80. I guess some optimization (or at least exploration of options) is in order before merging. (I'll push the version with in-place permutation to begin with.)

Any ideas? (One thing to test, I guess, is to reintroduce the trial-and-error approach, which at least is guaranteed to have a nonzero probability of termination, now… Maybe it's faster :D)

mlhetland commented:
Ah, I see `rem_edge!` uses `deleteat!` in a vector, which I guess would have a linear running time per use, which would be highly detrimental, here. Is it perhaps possible to instead modify the target node of the edge or the like?

mlhetland commented:
If I replace the use of `randperm!` with trial and error (i.e., just a `while true` loop, and `d = rand(1:n)`), things are closer to the original:

```
  memory estimate:  22.03 MiB
  allocs estimate:  313854
  --------------
  minimum time:     28.601 ms (0.00% GC)
  median time:      31.169 ms (3.72% GC)
  mean time:        30.290 ms (3.94% GC)
  maximum time:     35.549 ms (0.00% GC)
  --------------
  samples:          166
  evals/sample:     1
```

mlhetland commented:
For lower rewiring probabilities, the hit isn't so large (naturally enough, by the assumption that much of the extra cost is actually rewiring edges, rather than just adding them, as in the original). For example, if I instead use `β in 0.0:.1:.3`, I get the following:

**Original:**
```
  memory estimate:  11.89 MiB
  allocs estimate:  164863
  --------------
  minimum time:     9.779 ms (0.00% GC)
  median time:      9.984 ms (0.00% GC)
  mean time:        10.658 ms (6.16% GC)
  maximum time:     12.982 ms (19.58% GC)
  --------------
  samples:          470
  evals/sample:     1
```

**Modified:**
```
  memory estimate:  12.95 MiB
  allocs estimate:  180483
  --------------
  minimum time:     10.826 ms (0.00% GC)
  median time:      11.113 ms (0.00% GC)
  mean time:        11.831 ms (6.19% GC)
  maximum time:     14.160 ms (17.95% GC)
  --------------
  samples:          423
  evals/sample:     1
```

mlhetland commented:
I guess one could still use a streaming/reservoir sample to find a valid neighbor, to ensure a single pass (though then one would never use less than that), e.g., for cases where `k` is very high (with a low probability of randomly hitting a non-neighbor by trial and error). Doesn’t sound like a very reasonable tradeoff for normal uses, as `k` is normally (e.g., by Watts and Strogatz themselves) assumed to be much smaller than `n`.

mlhetland commented:
Just to clarify, the currently committed version is my suggested solution, i.e., ready for possible approval/merging :-)

(But I'm also certainly open to suggestions/request or the like.)

sbromberger commented:
Thanks for your patience and for your streamlining efforts. Will review now.

sbromberger commented on src/SimpleGraphs/generators/randgraphs.jl:
```diff
@@ -237,40 +253,54 @@ julia> watts_strogatz(10, 4, 0.3)
+        return is_directed ? complete_digraph(n) : complete_graph(n)
     end
+
+    g = is_directed ? SimpleDiGraph(n) : SimpleGraph(n)
```
Nice use of the ternary here.

sbromberger commented on src/SimpleGraphs/generators/randgraphs.jl:
```diff
@@ -237,40 +253,54 @@ julia> watts_strogatz(10, 4, 0.3)
-                end
-            end
+
+    target(s, i) = ((s + i - 1) % n) + 1
```
we generally don't define nested functions, but I think this might be ok. Could you add a comment here describing why this works?

mlhetland replied:
Sure! (Just reusing `target = ((s + i - 1) % n) + 1` from the original, really. Made it a function, since I'm using it twice.)

sbromberger replied:
Might want to put an `@inline` in there then just to help the compiler out.

sbromberger commented on src/SimpleGraphs/generators/randgraphs.jl:
```diff
@@ -237,40 +253,54 @@ julia> watts_strogatz(10, 4, 0.3)
+            d == t && break
+            add_edge!(g, s, d) && rem_edge!(g, s, t) && break
         end
+
```
don't need these blank lines.

sbromberger commented on src/SimpleGraphs/generators/randgraphs.jl:
```diff
@@ -237,40 +253,54 @@ julia> watts_strogatz(10, 4, 0.3)
     end
+
     return g
+
```
don't need blank lines here.

sbromberger reviewed (commented):
Thank you for this. A few comments inline, but one general one: would you mind adding more comments inline describing the main points?

mlhetland commented:
> Thanks for your patience and for your streamlining efforts.

No worries! Thanks for taking the time to look at this :-)

> Will review now.

Great! I'll have a look at the review.

mlhetland commented:
> Thank you for this. A few comments inline, but one general one: would you mind adding more comments inline describing the main points?

Sure! Just tried to follow the style of the original, which didn't have any comments (except for a TODO :-)). Can certainly add some!

mlhetland commented:
I've tried to address your concerns in the previous commit.

mlhetland commented:
Oh, and by the way, if there are minor stylistic issues where I diverge from the house style (e.g., with blank lines or the like), feel free to just edit the PR directly, if you wish. (That holds for other issues too, for that matter :-))

sbromberger commented on src/SimpleGraphs/generators/randgraphs.jl:
```diff
@@ -237,40 +253,66 @@ julia> watts_strogatz(10, 4, 0.3)
+            d == t && break             # Rewired to original target
+            if add_edge!(g, s, d)       # Was this valid (i.e., unconnected)?
+                rem_edge!(g, s, t)      # True rewiring: Delete original edge
+                break                   # We found a valid target
```
fantastic comments. Thank you.

sbromberger commented:
> if there are minor stylistic issues where I diverge from the house style (e.g., with blank lines or the like), feel free to just edit the PR directly, if you wish

I'm happy to do this but my experience is that folks who go through the process from start to finish tend to stick around longer, and their future PRs require substantially less review :)

Thanks for sticking with it. I've got one more suggestion and then I think we're ready to merge.

mlhetland commented:
Sure! I just meant I wouldn't object if you wanted to tweak something – but I'd be happy to finalize things :-) Added `@inline` now.

sbromberger commented:
Thanks for your help with this!
